### PR TITLE
[Feature] Ascend backend

### PR DESCRIPTION
**Pull request by lzhangzz**

### CODE REVIEW

**Comment by lvhan028** [comment]
please resolve conflicts and circleci issues by merging master

File: mmdeploy/backend/ascend/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+
+_from_acl_data_type = {0: torch.float32, 3: torch.int32, 9: torch.int64}
+
+ACL_MEMCPY_HOST_TO_HOST = 0
+ACL_MEMCPY_HOST_TO_DEVICE = 1
+ACL_MEMCPY_DEVICE_TO_HOST = 2
+ACL_MEMCPY_DEVICE_TO_DEVICE = 3
+
+
+class Error(Exception):
```
Comment (tpoisonooo) [inline mmdeploy/backend/ascend/wrapper.py:23]: It looks like a useless class.

**Review by tpoisonooo — COMMENTED** [review]
Would you please add a CI to build  mmdeploy SDK with ascend backend ?  I think compile demo do not need real atlas device.

File: csrc/mmdeploy/net/acl/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+project(mmdeploy_acl_net)
+
+if ("acl" IN_LIST MMDEPLOY_TARGET_BACKENDS)
+    message(STATUS "${ASCEND_TOOLKIT_HOME}")
+    mmdeploy_add_module(${PROJECT_NAME} acl_net.cpp)
+    target_include_directories(${PROJECT_NAME} PRIVATE
+        $<BUILD_INTERFACE:${ASCEND_TOOLKIT_HOME}/acllib/include>)
+    target_link_directories(${PROJECT_NAME} PUBLIC
+        # ${ASCEND_TOOLKIT_HOME}/runtime/lib64/stub
```
Comment (RunningLeon) [inline csrc/mmdeploy/net/acl/CMakeLists.txt:11]: may remove for no confusion

File: tools/test.py [hunk 3]
```diff
@@ -121,6 +121,7 @@ def main():
     is_device_cpu = (args.device == 'cpu')
     device_id = None if is_device_cpu else parse_device_id(args.device)
 
+    destroy_model = model.destroy
```
Comment (RunningLeon) [inline tools/test.py:124]: not all `End2EndModel` has `destroy` method.

File: csrc/mmdeploy/net/CMakeLists.txt [hunk 4]
```diff
@@ -26,5 +26,7 @@ if ("snpe" IN_LIST MMDEPLOY_TARGET_BACKENDS)
     add_subdirectory(snpe)
 endif ()
 
+add_subdirectory(acl)
```
Comment (lvhan028) [inline csrc/mmdeploy/net/CMakeLists.txt:29]:
```
if ("acl" IN_LIST MMDEPLOY_TARGET_BACKENDS)
    add_subdirectory(acl)
endif ()
```

File: docs/zh_cn/01-how-to-build/linux-x86_64.md [hunk 5]
```diff
@@ -235,6 +235,17 @@ export LD_LIBRARY_PATH=$Torch_DIR/lib:$LD_LIBRARY_PATH
   </code></pre>
   </td>
   </tr>
+  <tr>
+    <td>Ascend</td>
+    <td>CANN</td>
+    <td>
+    1. 按照 <a href="https://support.huawei.com/enterprise/zh/ascend-computing/cann-pid-251168373?category=installation-upgrade">官方指引</a> 安装 CANN 工具集.<br>
```
Comment (lvhan028) [inline docs/zh_cn/01-how-to-build/linux-x86_64.md:242]: 建议使用community版本的链接：https://www.hiascend.com/document/detail/zh/CANNCommunityEdition/60RC1alpha02/softwareinstall/instg/atlasdeploy_03_0002.html

File: configs/mmcls/classification_ascend_dynamic-224x224-224x224.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+_base_ = ['./classification_dynamic.py', '../_base_/backends/ascend.py']
+
+onnx_config = dict(input_shape=[224, 224])
+
+backend_config = dict(model_inputs=[
+    dict(
+        dynamic_batch_size=[1, 2, 4, 8],
```
Comment (lvhan028) [inline configs/mmcls/classification_ascend_dynamic-224x224-224x224.py:7]: Does it mean the dynamic range of batch dim is [1, 8]?

File: csrc/mmdeploy/core/device.h [hunk 7]
```diff
@@ -78,13 +78,13 @@ class Device {
 
   constexpr bool is_device() const noexcept { return platform_id() > 0; }
 
-  constexpr bool operator==(const Device &other) const noexcept {
-    return platform_id_ == other.platform_id_ && device_id_ == other.device_id_;
+  constexpr bool operator==(const Device& other) const noexcept {
+    return platform_id_ == other.platform_id_ & &device_id_ == other.device_id_;
```
Comment (lvhan028) [inline csrc/mmdeploy/core/device.h:82]: "& &" is weird.

File: csrc/mmdeploy/core/device.h [hunk 8]
```diff
@@ -78,13 +78,13 @@ class Device {
...
   }
 
-  constexpr bool operator!=(const Device &other) const noexcept { return !(*this == other); }
+  constexpr bool operator!=(const Device& other) const noexcept { return !(*this == other); }
 
-  constexpr explicit operator bool() const noexcept { return platform_id_ >= 0 && device_id_ >= 0; }
+  constexpr explicit operator bool() const noexcept { return platform_id_ >= 0 & &device_id_ >= 0; }
```
Comment (lvhan028) [inline csrc/mmdeploy/core/device.h:87]: strange '& &'

File: csrc/mmdeploy/core/device_impl.cpp [hunk 9]
```diff
@@ -321,6 +321,10 @@ int PlatformRegistry::Register(Creator creator) {
   return 0;
 }
 
+int PlatformRegistry::AddAlias(const char* name, const char* target) {
+  aliases_.emplace_back(name, target);
+}
```
Comment (lvhan028) [inline csrc/mmdeploy/core/device_impl.cpp:326]:
```
warning: no return statement in function returning non-void [-Wreturn-type]
```

File: csrc/mmdeploy/net/acl/acl_net.cpp [hunk 10]
```diff
@@ -0,0 +1,656 @@
...
+#include "mmdeploy/core/logger.h"
+#include "mmdeploy/core/model.h"
+#include "mmdeploy/core/utils/formatter.h"
+
+std::ostream& operator<<(std::ostream& os, const aclmdlIODims& dims) {
+  os << dims.name << " [";
+  for (int i = 0; i < dims.dimCount; ++i) {
+    os << (i ? ", " : "") << dims.dims[i];
+  }
+  os << "]";
```
Comment (lvhan028) [inline csrc/mmdeploy/net/acl/acl_net.cpp:14]:
```
 warning: no return statement in function returning non-void [-Wreturn-type]
```

File: csrc/mmdeploy/net/acl/acl_net.cpp [hunk 11]
```diff
@@ -0,0 +1,656 @@
...
+}
+
+std::ostream& operator<<(std::ostream& os, const aclmdlBatch& batch) {
+  os << "batch [";
+  for (int i = 0; i < batch.batchCount; ++i) {
+    os << (i ? ", " : "") << batch.batch[i];
+  }
+  os << "]";
```
Comment (lvhan028) [inline csrc/mmdeploy/net/acl/acl_net.cpp:22]:
```
 warning: no return statement in function returning non-void [-Wreturn-type]
```

File: csrc/mmdeploy/net/acl/acl_net.cpp [hunk 12]
```diff
@@ -0,0 +1,656 @@
...
+}
+
+std::ostream& operator<<(std::ostream& os, const aclmdlHW& hw) {
+  os << "HW [";
+  for (int i = 0; i < hw.hwCount; ++i) {
+    os << (i ? ", " : "") << "(" << hw.hw[i][0] << ", " << hw.hw[i][1] << ")";
+  }
+  os << "]";
```
Comment (lvhan028) [inline csrc/mmdeploy/net/acl/acl_net.cpp:30]:
```
 warning: no return statement in function returning non-void [-Wreturn-type]
```

File: docs/en/01-how-to-build/linux-x86_64.md [hunk 13]
```diff
@@ -238,6 +238,17 @@ export LD_LIBRARY_PATH=$Torch_DIR/lib:$LD_LIBRARY_PATH
...
     </td>
   </tr>
+  <tr>
+    <td>Ascend</td>
+    <td>CANN</td>
+    <td>
+    1. Install CANN follow <a href="https://www.hiascend.com/document/detail/en/CANNCommunityEdition/51RC1alphaX/softwareinstall/instg/atlasdeploy_03_0002.html">official guide</a>.<br>
+    2. Setup environment
+   <pre><code>
+export ASCEND_TOOLKIT_HOME="/usr/local/Ascend/ascend-toolkit/latest"
```
Comment (lvhan028) [inline docs/en/01-how-to-build/linux-x86_64.md:248]: Should we use `source set_env.sh` instead?

File: mmdeploy/backend/ascend/utils.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import math
+import os.path as osp
+
+from mmdeploy.utils import get_root_logger
+
+
+def update_sdk_pipeline(work_dir: str):
```
Comment (lvhan028) [inline mmdeploy/backend/ascend/utils.py:8]: What's this function for, especially considering dynamic (h, w) is supported by `detection_ascend_dynamic_800x1344`?

File: mmdeploy/backend/ascend/onnx2ascend.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+    input_format = 'ND' if 'dynamic_dims' in model_inputs else 'NCHW'
+
+    args = [
+        f'--model={onnx_path}', '--framework=5', f'--output={output_path}',
+        '--soc_version=Ascend310', f'--input_format={input_format}',
+        f'--input_shape={input_shapes}'
+    ]
+
+    if 'dynamic_batch_size' in model_inputs:
```
Comment (lvhan028) [inline mmdeploy/backend/ascend/onnx2ascend.py:67]: So `dynamic_batch_size`, `dynamic_image_size` and `dynamic_dims` are exclusive from each other? @grimoire

File: csrc/mmdeploy/net/acl/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+if ("acl" IN_LIST MMDEPLOY_TARGET_BACKENDS)
+    if (NOT DEFINED ASCEND_TOOLKIT_HOME)
+        set(ASCEND_TOOLKIT_HOME $ENV{ASCEND_TOOLKIT_HOME})
+    endif ()
+    mmdeploy_add_module(${PROJECT_NAME} acl_net.cpp)
+    target_include_directories(${PROJECT_NAME} PRIVATE
+        $<BUILD_INTERFACE:${ASCEND_TOOLKIT_HOME}/runtime/include>)
+    target_link_libraries(${PROJECT_NAME} PRIVATE
+        $<BUILD_INTERFACE:${ASCEND_TOOLKIT_HOME}/runtime/lib64/stub/libascendcl.so>)
```
Comment (lvhan028) [inline csrc/mmdeploy/net/acl/CMakeLists.txt:13]: Why we link the stub lib?

Comment (lzhangzz) [reply csrc/mmdeploy/net/acl/CMakeLists.txt:13]: When in a CI build or cross compiling, you don't have the driver libraries needed by non-stub libraries.

**Comment by lvhan028** [comment]
@grimoire May update `check_env.py`

File: configs/mmcls/classification_ascend_dynamic-224x224-224x224.py (hunk 6 above)
Comment (lzhangzz) [reply configs/mmcls/classification_ascend_dynamic-224x224-224x224.py:7]: Actually, it's just 1, 2, 4 or 8, any other number will trigger an error in their API. We choose the smallest one suitable for input's batch size in our wrapper classes.

File: csrc/mmdeploy/net/acl/CMakeLists.txt (hunk 16 above)
Comment (lzhangzz) [reply csrc/mmdeploy/net/acl/CMakeLists.txt:13]: The first one is real and the second one is the stub.

File: mmdeploy/backend/ascend/wrapper.py (hunk 1 above)
Comment (lzhangzz) [reply mmdeploy/backend/ascend/wrapper.py:23]: It's used to allow user code like
```python
try:
   # do something
except OSError as e:
   # report os error
except mmdeploy.backend.ascend.Error as e:
   # report ascend error
```

**Comment by lzhangzz** [comment]
> Would you please add a CI to build mmdeploy SDK with ascend backend ? I think compile demo do not need real atlas device.

done

File: tools/test.py (hunk 3 above)
Comment (lzhangzz) [reply tools/test.py:122]: A `destroy` method is also added in `BaseBackendModel`

https://github.com/open-mmlab/mmdeploy/blob/1d2ca0d232a2a9e52b9eb59934e99b83f2fe8254/mmdeploy/codebase/base/backend_model.py#L122-L124

File: mmdeploy/backend/ascend/utils.py (hunk 14 above)
Comment (grimoire) [reply mmdeploy/backend/ascend/utils.py:8]: Ascend does not support dynamic shape for any shape.
We can export two shapes [800x1344] & [1344x800], which need a special way to pad the input. This function is used to update the pad mode to support such transform.

**Review by lvhan028 — APPROVED** [review]

**Review by tpoisonooo — APPROVED** [review]

**Comment by tpoisonooo** [comment]
`cmake_option.md` 也应该加一行。

File: mmdeploy/backend/ascend/wrapper.py [hunk 17]
```diff
@@ -0,0 +1,593 @@
...
+            elif ret == 100037:  # ACL_ERROR_REPEAT_FINALIZE
+                pass
+            else:
+                _check(ret, 'acl.finalize')
+        self._active = False
+
+
+_is_torch_npu_available = False
+
+if os.environ.get('MMDEPLOY_USE_TORCH_NPU'):
```
Comment (RunningLeon) [inline mmdeploy/backend/ascend/wrapper.py:269]: should we mention this env var in the doc?

**Review by RunningLeon — APPROVED** [review]
LGTM

File: mmdeploy/backend/ascend/wrapper.py (hunk 17 above)
Comment (lzhangzz) [reply mmdeploy/backend/ascend/wrapper.py:269]: Let's fix documentation later in another PR.

**Pull request merged**